pkg/component/server/konnectivity.go

```diff
@@ -164,95 +199,48 @@ func (k *Konnectivity) writeKonnectivityAgent() error {
 	return nil
 }
 
-func (k *Konnectivity) runLease() {
-	k.leaseCtx, k.leaseCancel = context.WithCancel(context.Background())
-	go func(ctx context.Context) {
-		logrus.Infof("starting %v lease watcher", serviceName)
-		leaseLock, err := k.newLeaseLock()
-		if err != nil {
-			logrus.Error(err)
-		}
-		k.leaseLock = leaseLock
-		for {
-			select {
-			case <-ctx.Done():
-				logrus.Debugf("stopping lease watcher for %v", serviceName)
-				return
-			default:
-				k.leaseLock.LeaseRunner(context.Background())
+func (k *Konnectivity) runLeaseCounter() {
+
+	logrus.Infof("starting to count controller lease holders every 10 secs")
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-k.stopCtx.Done():
+			return
+		case <-ticker.C:
+			count, err := k.countLeaseHolders()
+			if err != nil {
+				logrus.Errorf("failed to count controller leases: %s", err)
+				continue
 			}
+			k.serverCountChan <- count
 		}
-	}(k.leaseCtx)
-
-	go func(ctx context.Context) {
-		logrus.Infof("watching %v lease holders", serviceName)
-		ticker := time.NewTicker(30 * time.Second)
-		defer ticker.Stop()
-		for {
-			select {
-			case <-ctx.Done():
-				logrus.Debugf("stopping lease holder count for %v", serviceName)
-				return
-			case <-ticker.C:
-				observedLeaseHolders := k.leaseLock.CountValidLeaseHolders(context.Background())
-				if observedLeaseHolders != k.serverCount {
-					logrus.Debugf("change in %v lease holders detected. refreshing service.", serviceName)
-					k.serverCount = observedLeaseHolders
-
-					// restarting service
-					if err := k.restartService(); err != nil {
-						logrus.Errorf("failed to restart %v: %v", serviceName, err)
-					}
-				}
-			}
-			logrus.Debugf("found %v lease holders for %v", k.serverCount, serviceName)
-		}
-	}(k.leaseCtx)
+	}
 }
 
-func (k *Konnectivity) newLeaseLock() (*kubernetes.LeaseLock, error) {
+func (k *Konnectivity) countLeaseHolders() (int, error) {
 	client, err := k.KubeClientFactory.GetClient()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get lease client: %v", err)
+		return 0, err
 	}
-
-	holderIdentity, err := os.Hostname()
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

Comment:
```suggestion
	ctx, cancel := context.WithTimeout(k.stopCtx, 5*time.Second)
```
So we abort on Stop().